### ISSUE TITLE
box type tests and encrypted_nonce

### ISSUE THREAD

**Issue by keks** [ssbc/envelope-spec#5]
I was about to build a test runner for this in go, but it's currently not compatible with the function signatures I use. Specifically, in Go I use a list of info-strings that ties the key to the particular context. The test JSON and JS implementation operate on the `external_nonce` directly.

I am convinced we should pass in the logic values that are included in the context, instead of the result of the computation. First, in the JS implementation the function for computing the `external_nonce` is just completely missing. Second, the meaning and encoding of the values, and the compuation performed to get the `external_nonce` need to be tested as well.

I guess this issue ties in with #4.

I understand you may have discussed this and I'm missing something, I'd be interested in what that is.

**Comment by mixmix** [comment]
yeah @keks ,  we should definitely discuss!

As I've said this is a first pass and I expect it to change. I wanted a starting point to force us to discuss and run shared tests.

Past thinking (if I remember right):
1. **I didn't include derivation of `external_nonce` because it wasn't yet clearly defined**, so I made it arbitrary (assuming a 32 byte buffer). Very keen to specify it and move the calculation to the safest + most appropriate place
    - I think forcing passing in "previous" to box/ unbox could really lock in the importance of the feed/  chain context that box2 is expected to behave in
    - alternatively it could bind it to a context it doesn't need to
2. **When I came to filling the `key_slot`s, I was unclear on how to do it**, so I just used the same function as `hkdf.expand` (why not use the same hash thing everywhere!)
    - for the `label` part I knew we needed something unique to this `msg_key` and if `external_nonce` was derived from `previous` ... maybe we could just use that
    - very happy for this to change

I've only lightly discussed this with @dominictarr . The only thing we strongly agreed on was that having some test frame would allow us to iterate more clearly to our final spec, and I should start somewhere!
